pkg/node/manager/manager.go

```diff
@@ -333,134 +333,153 @@ func (m *manager) legacyNodeIpBehavior() bool {
-		// encapsulation. In encryption case we also want to use vxlan device
-		// to create symmetric traffic when sending nodeIP->pod and pod->nodeIP.
-		if address.Type == addressing.NodeCiliumInternalIP || m.conf.EncryptionEnabled() ||
-			option.Config.EnableHostFirewall || option.Config.JoinCluster {
-			tunnelIP = nodeIP
-		}
-
 		if option.Config.NodeIpsetNeeded() && address.Type == addressing.NodeInternalIP {
 			iptables.AddToNodeIpset(address.IP)
 		}
 
-		if skipIPCache(address) {
+		if m.nodeAddressSkipsIPCache(address) {
 			continue
 		}
 
-		// If we are doing encryption, but not node based encryption, then do not
-		// add a key to the nodeIPs so that we avoid a trip through stack and attempting
-		// to encrypt something we know does not have an encryption policy installed
-		// in the datapath. By setting key=0 and tunnelIP this will result in traffic
-		// being sent unencrypted over overlay device.
-		if (!m.conf.NodeEncryptionEnabled() &&
-			(address.Type == addressing.NodeExternalIP || address.Type == addressing.NodeInternalIP)) ||
-			// Also ignore any remote node's key if the local node opted to not perform
-			// node-to-node encryption
-			node.GetOptOutNodeEncryption() {
-			key = 0
+		var tunnelIP netip.Addr
+		if m.nodeAddressHasTunnelIP(address) {
+			tunnelIP = nodeIP
 		}
 
-		var prefix netip.Prefix
-		if v4 := address.IP.To4(); v4 != nil {
-			prefix = ip.IPToNetPrefix(v4)
-		} else {
-			prefix = ip.IPToNetPrefix(address.IP.To16())
+		var key uint8
+		if m.nodeAddressHasEncryptKey(address) {
+			key = n.EncryptionKey
 		}
-		ipAddrStr := prefix.String()
-		_, err := m.ipcache.Upsert(ipAddrStr, tunnelIP, key, nil, ipcache.Identity{
-			ID:     remoteHostIdentity,
-			Source: n.Source,
-		})
-		resource := ipcacheTypes.NewResourceID(ipcacheTypes.ResourceKindNode, "", n.Name)
-		m.upsertIntoIDMD(prefix, remoteHostIdentity, resource)
-
-		// Upsert() will return true if the ipcache entry is owned by
-		// the source of the node update that triggered this node
-		// update (kvstore, k8s, ...) The datapath is only updated if
-		// that source of truth is updated.
+
+		prefix := ip.IPToNetPrefix(address.IP)
+		// We expect the node manager to have a source of either Kubernetes,
+		// CustomResource, or KVStore. Prioritize the KVStore source over the
+		// rest as it is the strongest source, i.e. only trigger datapath
+		// updates if the information we receive takes priority.
+		//
 		// The only exception are kube-apiserver entries. In that case,
 		// we still want to inform subscribers about changes in auxiliary
 		// data such as for example the health endpoint.
-		overwriteErr := &ipcache.ErrOverwrite{
-			ExistingSrc: source.KubeAPIServer,
-			NewSrc:      n.Source,
-		}
-		if err != nil && !errors.Is(err, overwriteErr) {
+		existing := m.ipcache.GetMetadataByPrefix(prefix).Source()
+		overwrite := source.AllowOverwrite(existing, n.Source)
+		if !overwrite && existing != source.KubeAPIServer {
```

Comment:
Ultimately, checking for `existing != source.KubeAPIServer` was missing from the previous revision that was causing CI kind failures

Reply:
cc @gandro @joestringer, this was needed to resolve the encryption issue on upgrade on kind clusters (with tunneling), part 2/2.

Reply:
Wrote down my thoughts on this here: https://github.com/cilium/cilium/pull/23208#issuecomment-1543570384

I don't think we should address this in this PR (it's already hard enough keeping track of the changes), but I do think we should think about this as part of the refactor, because I really think kube-apiserver is not that special. Future sources (e.g. label-sources) might suffer a similar problem.

Reply:
Added it to https://github.com/cilium/cilium/issues/21142, agreed.